pkg/cmd/pulumi/util_remote.go

```diff
@@ -0,0 +1,315 @@
+			"NAME=secretvalue (e.g. `--remote-env FOO=secret`)")
+	cmd.PersistentFlags().StringArrayVar(
+		&r.preRunCommands, "remote-pre-run-command", []string{},
+		"[EXPERIMENTAL] Commands to run before the remote operation")
+	cmd.PersistentFlags().StringVar(
+		&r.gitBranch, "remote-git-branch", "",
+		"[EXPERIMENTAL] Git branch to deploy; this is mutually exclusive with --remote-git-branch; "+
+			"either value needs to be specified")
+	cmd.PersistentFlags().StringVar(
+		&r.gitCommit, "remote-git-commit", "",
+		"[EXPERIMENTAL] Git commit hash of the commit to deploy (if used, HEAD will be in detached mode); "+
+			"this is mutually exclusive with --remote-git-branch; either value needs to be specified")
+	cmd.PersistentFlags().StringVar(
+		&r.gitRepoDir, "remote-git-repo-dir", "",
+		"[EXPERIMENTAL] The directory to work from in the project's source repository "+
+			"where Pulumi.yaml is located; used when Pulumi.yaml is not in the project source root")
+	cmd.PersistentFlags().StringVar(
+		&r.gitAuthAccessToken, "remote-git-auth-access-token", "",
+		"[EXPERIMENTAL] Git personal access token")
+	cmd.PersistentFlags().StringVar(
+		&r.gitAuthSSHPrivateKey, "remote-git-auth-ssh-private-key", "",
+		"[EXPERIMENTAL] Git SSH private key; use --remote-git-auth-password for the password, if needed")
+	cmd.PersistentFlags().StringVar(
+		&r.gitAuthSSHPrivateKeyPath, "remote-git-auth-ssh-private-key-path", "",
+		"[EXPERIMENTAL] Git SSH private key path; use --remote-git-auth-password for the password, if needed")
+	cmd.PersistentFlags().StringVar(
+		&r.gitAuthPassword, "remote-git-auth-password", "",
+		"[EXPERIMENTAL] Git password; for use with username or with an SSH private key")
+	cmd.PersistentFlags().StringVar(
+		&r.gitAuthUsername, "remote-git-auth-username", "",
+		"[EXPERIMENTAL] Git username")
+}
+
+// runDeployment kicks off a remote deployment.
+func runDeployment(ctx context.Context, opts display.Options, operation apitype.PulumiOperation, stack, url string,
+	args RemoteArgs) result.Result {
+
+	b, err := currentBackend(ctx, opts)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	// Ensure the cloud backend is being used.
+	cb, isCloud := b.(httpstate.Backend)
+	if !isCloud {
+		return result.FromError(errors.New("the Pulumi service backend must be used for remote operations; " +
+			"use `pulumi login` without arguments to log into the Pulumi service backend"))
+	}
+
+	stackRef, err := b.ParseStackReference(stack)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	if url == "" {
+		return result.FromError(errors.New("Git repo URL most be specified"))
+	}
+	if args.gitCommit != "" && args.gitBranch != "" {
+		return result.FromError(errors.New("Git commit and branch cannot both be specified"))
+	}
+	if args.gitCommit == "" && args.gitBranch == "" {
+		return result.FromError(errors.New("at least Git commit or branch are required"))
+	}
```

Comment:
"at least" implies 1 or more. Could we change the bottom message to something like
```
either "--remote-git-branch" or "--remote-git-commit" is required
```